src/ng-xform/select-field/select-field.component.ts

```diff
@@ -1,32 +1,114 @@
+import 'rxjs/add/operator/switchMap';
+import 'rxjs/add/operator/distinctUntilChanged';
+import 'rxjs/add/operator/debounceTime';
 
+import { BaseDynamicFieldComponent } from './../field-components/base-dynamic-field.component';
 import { SelectField } from '../fields';
-import { BaseDynamicFieldComponent } from '../field-components/base-dynamic-field.component';
 
 
 /**
  * Component to generate a bootstrap form field of general type
  *
- * :inputId: the id of the field in the FormGroup
- * :inputType: type of the input field
- * :caption: Title of the field
+ * :editing: Flag to control component state
  * :form: FormGroup containing the field
- * :validator: external function that validates the data in the field
- * :errorMessage: message to be shown when the validation fails
+ * :field: Intance of field configurations
  */
 @Component({
   selector: 'ng-xform-select-field',
   templateUrl: './select-field.component.html',
+  providers: [{
+    provide: NG_VALUE_ACCESSOR,
+    useExisting: SelectFieldComponent,
+    multi: true
+  }],
 })
-export class SelectFieldComponent extends BaseDynamicFieldComponent<SelectField> implements OnInit {
+export class SelectFieldComponent extends BaseDynamicFieldComponent<SelectField> implements ControlValueAccessor, OnInit, AfterViewInit {
 
-  @Input() errorMessage: string;
+  @ViewChild(NgSelectComponent) select: NgSelectComponent;
+  viewModel = new BehaviorSubject<any>(null);
 
   optionValues: any[] = [];
+  optionLabel = '-';
+  typeahead: EventEmitter<string>;
+
+  _onChange: any = (value: any) => { };
+  _onTouched: any = () => { };
 
   ngOnInit() {
     super.ngOnInit();
+    this.config();
+  }
+
+  ngAfterViewInit() {
+    this.select.registerOnChange((value: any) => {
+      this.updateOptionLabel();
+      this._onChange(value);
+    });
+    this.select.registerOnTouched(this._onTouched);
+    this.viewModel.subscribe((value: any) => {
+      if (this.field.searchByValueAttributeHandler) {
+        this.field.searchByValueAttributeHandler(value).subscribe((val: any) => {
+          this.optionValues = [val];
+          this.select.writeValue(value);
+          this.updateOptionLabel();
+        });
+        return;
+      }
+      this.select.writeValue(value);
+      this.updateOptionLabel();
+    })
+  }
+
+  writeValue(obj: any): void {
+    this.viewModel.next(obj);
+  }
+
+  registerOnChange(fn: any): void {
+    this._onChange = fn;
+  }
+  registerOnTouched(fn: any): void {
+    this._onTouched = fn;
+  }
+  setDisabledState?(isDisabled: boolean): void {
+    this.select.setDisabledState(isDisabled);
+  }
+
+  updateOptionLabel() {
+    setTimeout(() => {
+      this.optionLabel = this.select.selectedItems.map(item => item.label).join(this.field.separator);
+    }, 0);
+  }
+
+  config() {
+    if (this.field.searchHandler) {
+      this.typeahead = new EventEmitter<string>();
+      this.field.searchable = true;
+      let searchHandler = this.typeahead.asObservable();
```

Comment:
What about moving lines 92 to 101 to a new method called `getTypeaheadWithDistinctAndDebounce`? Would make the code a lot more clear